src/overseer/core.clj

```diff
@@ -1,97 +1,62 @@
-           to-visit #{job-ent-id}]
-      (if (empty? to-visit)
-        (set (map #(:job/id (d/entity db %)) all-dependents))
-        (let [ent (first to-visit)
-              dependents (ent-dependents db ent)
-              all-dependents' (set/union all-dependents dependents)
-              visited' (conj visited ent)
-              to-visit' (set/difference
-                          (set/union to-visit dependents)
-                          visited')]
-          (recur all-dependents' visited' to-visit'))))))
-
-(defn status-txn
-  "Construct a single job update status txn"
-  [{:keys [overseer/status overseer/failure]} job-id]
-  (let [base-txn {:db/id [:job/id job-id]
-                  :job/status status}]
-    (if failure
-      (assoc base-txn :job/failure (pr-str failure))
-      base-txn)))
-
-(defn update-job-status-txns
-  "Construct a seq of txns for updating a job's status
-   If job was aborted, then also abort all its dependents"
-  [db job-id {:keys [overseer/status] :as status-map}]
-  (let [job-ids (if (= :aborted status)
-                  (cons job-id (transitive-dependents db job-id))
-                  [job-id])]
-    (map (partial status-txn status-map) job-ids)))
+(defprotocol Store
+  (job-info [this job-id])
+
+  (reserve-job [this job-id])
+  (finish-job [this job-id])
+  (fail-job [this job-id failure])
+  (heartbeat-job [this job-id])
+  (abort-job [this job-id])
+  (reset-job [this job-id])
+
+  (jobs-ready [this])
+  (jobs-dead [this thresh])
+
+  (transact-graph [this graph]))
+
+(defn squuid []
+  (let [uuid (java.util.UUID/randomUUID)
+        time (System/currentTimeMillis)
+        secs (quot time 1000)
+        lsb (.getLeastSignificantBits uuid)
+        msb (.getMostSignificantBits uuid)
+        timed-msb (bit-or (bit-shift-left secs 32)
+                          (bit-and 0x00000000ffffffff msb))]
+    (java.util.UUID. timed-msb lsb)))
+
+(defn missing-handlers [handlers keyword-adjacency-list]
+  (-> (loom.graph/nodes (loom.graph/digraph keyword-adjacency-list))
+      (set/difference (keys handlers))))
+
+(defn job-graph [keyword-adjacency-list tx]
+  (let [keyword-graph (loom.graph/digraph keyword-adjacency-list)
+
+        job-ids-by-job-type ; {:job-type-foo #uuid "abc123..."}
+        (std/zipmap-seq identity (fn [_] (squuid)) (loom.graph/nodes keyword-graph))
+
+        job-type->job-map
+        (fn [job-type]
+          (merge
+            {:job/id (job-ids-by-job-type job-type)
+             :job/status :unstarted
+             :job/type job-type}
+            tx))]
+    (loom.derived/mapped-by job-type->job-map keyword-graph)))
+
+(defn ready-job-info
+  "Return the set of job entities that are ready to execute,
+   filtering to those defined in job-handlers (this allows different
+   nodes to solely execute certain types of jobs, if desired)"
+  [store job-handlers]
+  (->> (jobs-ready store)
+       (map (partial job-info store))
+       (filter (comp job-handlers :job/type))
```

Comment:
As an aside, I have always had to do a double-take on this line; I really think it may have been clearer as 

```clj
(filter #(contains? job-handlers (:job/type %)))
```